### PR TITLE
dvclive 1.0

### PR DESCRIPTION
**Pull request by daavoo**
Changes from https://github.com/iterative/dvclive/issues/223

### CODE REVIEW

**Comment by dberenbaum** [comment]
@daavoo Do you want a review or should I wait?

File: content/docs/dvclive/api-reference/live/log_image.md [hunk 1]
```diff
@@ -30,44 +30,27 @@ Supported values for `val` are:
   [PIL.Image.fromarray](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.fromarray))
 - A `PIL.Image` instance.
 
-The images will be saved in `{Live.dir}/images/{name}`:
+The images will be saved in `{Live.plots_dir}/images/{name}`:
 
 ```dvc
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_image.md:35]: 💅🏼 Good opportunity to update al these to ` ```cli `

File: content/docs/dvclive/api-reference/live/log_image.md [hunk 2]
```diff
@@ -30,44 +30,27 @@ Supported values for `val` are:
...
 $ tree
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_image.md:36]: 💅🏼 Also, no need to show the `tree` command IMO (not available on Windows by default, maybe same with Mac). It's clear that this is a folder structure, I think.

File: content/docs/dvclive/api-reference/live/log_image.md [hunk 3]
```diff
@@ -30,44 +30,27 @@ Supported values for `val` are:
...
-$ tree
-└── dvclive
-    └── images
-        ├── 0
-        │   ├── numpy.png
-        │   └── pil.png
-        └── 1
-            ├── numpy.png
-            └── pil.png
+dvc plots diff dvclive/plots
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_image.md:49]: These need `$` though
```suggestion
$ dvc plots diff dvclive/plots
```

File: content/docs/dvclive/api-reference/live/log_metric.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
+# Live.log_metric()
+
+```py
+ def log_metric(name: str, val: float):
+```
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_metric.md:5]: Minor since it's just one file but is it possible to have the Git diff recon that this is the old content/docs/dvclive/api-reference/live/log.md ? Would be easier to review.

> May need to drop this one from the repo and re-add the previous file, `git commit`, and then `git mv` it to this.

File: content/docs/dvclive/api-reference/live/log_param.md [hunk 5]
```diff
@@ -30,6 +28,16 @@ lr: 1e-3
 n_iter: 50
 ```
 
+<admon type="tip">
+
+The logged params can be visualized with `dvc params`:
+
+```dvc
+dvc params diff dvclive/params.yaml
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_param.md:36]: (There's more of these prob.)
````suggestion
```cli
$ dvc params diff dvclive/params.yaml
````

File: content/docs/dvclive/api-reference/live/log_image.md [hunk 6]
```diff
@@ -26,48 +26,30 @@ live.log_image("pil.png", img_pil)
...
   [PIL.Image.fromarray](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.fromarray))
 - A `PIL.Image` instance.
 
-The images will be saved in `{Live.dir}/images/{name}`:
+The images will be saved in `{Live.plots_dir}/images/{name}`:
 
-```dvc
-$ tree
+```
 └── dvclive
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_image.md:36]: 💅🏼 
```suggestion
dvclive
```
May need updating the lines below too.

File: content/docs/dvclive/api-reference/live/log_param.md [hunk 7]
```diff
@@ -24,12 +22,22 @@ live.log_param("n_iter", 50)
 On each `live.log_param(name, val)` call, DVCLive will write the `name`
 parameter to `{Live.dir}/params.yaml` with the corresponding `val`:
 
-```dvc
+```cli
 $ cat dvclive/params.yaml
 lr: 1e-3
 n_iter: 50
 ```
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_param.md:29]: Minor again but I incline to use YAML blocks for these, e.g.:

````suggestion
```yaml
lr: 1e-3
n_iter: 50
```
````

You'd just have to make sure it's mentioned shortly after that we're looking at dvclive/params.yaml .

Comment (jorgeorpinel) [reply content/docs/dvclive/api-reference/live/log_param.md:29]: I mean shortly before.

File: content/docs/dvclive/outputs.md [hunk 8]
```diff
@@ -1,19 +1,19 @@
 # Output Folder Structure
 
-DVCLive will store the logged data under the folder (`path`) passed to
+DVCLive will store the logged data under the directory (`path`) passed to
```
Comment (jorgeorpinel) [inline content/docs/dvclive/outputs.md:3]: Call the doc "Directory Structure" ?

File: content/docs/dvclive/api-reference/live/log_metric.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/log_metric.md:5]: No intro text here? Used to be 

> Logs the given scalar `val` associating it with the given `name`.

**Review by jorgeorpinel — COMMENTED** [review]
Looking good so far. Probably mergeable already 👍🏼

File: content/docs/dvclive/get-started.md [hunk 10]
```diff
@@ -89,31 +82,33 @@ more details and other supported frameworks.
...
 would usually do:
 
-```dvc
+```cli
 $ python train.py
 ```
 
-DVCLive will generate a [report](/doc/dvclive/outputs#report) in
+DVCLive will generate a [report](/doc/dvclive/api-reference/live/make_report) in
 `dvclive/report.html` containing all the logged data. It will be automatically
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:90]: Definitely not a blocker, but do we have an issue for showing other report types (including vs code and studio) here? I know @shcheklein mentioned it but not sure we noted it anywhere in GitHub.

File: content/docs/dvclive/get-started.md [hunk 11]
```diff
@@ -157,7 +152,7 @@ Running stage 'train':
 ### Compare and Visualize DVC Experiments
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:152]: Another non-blocking follow up: should we make these tabs? It's a bit long and feels like a lot of space to just show different interfaces for the experiments table.

File: content/docs/dvclive/how-it-works.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+The contents of the directory will depend on the methods used:
+
+| Method                                                                    | Writes to                  |
+| ------------------------------------------------------------------------- | -------------------------- |
+| [live.log_metric](/doc/dvclive/api-reference/live/log_metric)             | `dvclive/plots/metrics`    |
+| [live.log_image](/doc/dvclive/api-reference/live/log_image)               | `dvclive/plots/images`     |
+| [live.log_param](/doc/dvclive/api-reference/live/log_param)               | `dvclive/params.yaml`      |
+| [live.log_sklearn_plot](/doc/dvclive/api-reference/live/log_sklearn_plot) | `dvclive/plots/sklearn`    |
+| [live.make_report](/doc/dvclive/api-reference/live/make_report)           | `dvclive/report.{md/html}` |
+| [live.make_summary](/doc/dvclive/api-reference/live/make_summary)         | `dvclive/metrics.json`     |
```
Comment (dberenbaum) [inline content/docs/dvclive/how-it-works.md:16]: Not blocking: somewhere on this page, it could help to mention these are called by `next_step`.

File: content/docs/dvclive/api-reference/index.md [hunk 13]
```diff
@@ -25,6 +25,12 @@ from dvclive import Live
 live = Live()
 ```
 
+Or use as a context manager:
+
+```python
+with Live() as live:
+```
+
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/index.md:33]: Not blocking (but probably would block if I wasn't trying to get this out today): can we remove the suggested usage without the context manager? It could result in unexpected behavior.

IMO there are 2 categories of people who will use it without the context manager:
1. People who can figure out themselves that they can omit it and understand the consequences of doing so (they don't need this documented).
2. People who won't be able to figure out this pattern for themselves but may also be surprised at the consequences (they probably should use the context manager).

File: content/docs/dvclive/api-reference/index.md [hunk 14]
```diff
@@ -86,7 +92,14 @@ See `Live.log_plot()`.
...
 
-See `Live.next_step()` and `Live.set_step()` for details.
+See `Live.next_step()`.
+
+Under the hood, `Live.next_step()` calls `Live.make_summary()` and
+`Live.make_report()`.
+
+If you want to decouple the `step` update from the rest of the calls, you can
+manually modify the `Live.step` property and call `Live.make_summary()` /
+`Live.make_report()`.
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/index.md:102]: Non-blocking suggested rewording:

```suggestion
If you want to update the `step` to a different number or don't want to call
`Live.make_summary()` / `Live.make_report()`, you can manually modify the
`Live.step` property (for example, `live.step = 1`).
```

Also, why no mention of `make_checkpoint()`?

File: content/docs/dvclive/get-started.md [hunk 15]
```diff
@@ -64,18 +58,17 @@ trainer.fit(model)
...
-        live.log(metric_name, value)
+    for epoch in range(NUM_EPOCHS):
+        train_model(...)
+        metrics = evaluate_model(...)
+        for metric_name, value in metrics.items():
+            live.log_metric(metric_name, value)
+        live.next_step()
 
-    live.next_step()
+    live.summary["additional_metric"] = post_train_evaluate(...)
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/docs/dvclive/get-started.md:71]: Can we move it to `api-reference/index.md` (and maybe add a short explanation of the summary there)?

File: content/docs/dvclive/api-reference/live/index.md [hunk 16]
```diff
@@ -21,59 +21,71 @@ from dvclive import Live
...
 
+Or use as a context manager:
+
+```py
+from dvclive import Live
+
+with Live() as live:
+    ...
+```
+
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/index.md:32]: Non-blocking: let's only keep the context manager (see https://github.com/iterative/dvc.org/pull/4085/files#r1014164477).

File: content/docs/dvclive/api-reference/index.md (hunk 13 above)
Comment (dberenbaum) [reply content/docs/dvclive/api-reference/index.md:33]: Also, can we avoid mentioning `context manager`? I think it's too technical.

File: content/docs/dvclive/api-reference/live/index.md [hunk 17]
```diff
@@ -21,59 +21,71 @@ from dvclive import Live
...
 
-`Live()` will remove all existing DVCLive related files under `path` unless
+`Live()` will remove all existing DVCLive related files under `dir` unless
 `resume=True`.
 
 </admon>
 
-## Attributes
+You can use `Live()` as a context manager. When exiting the context manager,
+`Live.make_summary()` and `Live.make_report()` will be called.
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/index.md:46]:
```suggestion
`with Live() as live:` will call `Live.make_summary()`
and `Live.make_report()` on exit to update the summary
and report with everything logged during the final step.
```

Not blocking

File: content/docs/dvclive/api-reference/live/log_metric.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+```
+├── dvclive
+│   └── plots
+│       └── metrics
+│           ├── train
+│           │   └── loss.tsv
+│           └── val
+│               └── loss.tsv
+└── dvclive.json
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_metric.md:31]: **Blocking**: this looks like it's based on the old output format. Also, given the example above, there shouldn't be any summary file, right?

Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_metric.md:31]:
```suggestion
└── dvclive
    └── plots
        └── metrics
            ├── train
            │   └── loss.tsv
            └── val
                └── loss.tsv
```

File: content/docs/dvclive/api-reference/live/log_metric.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+```py
+ def log_metric(name: str, val: float):
+```
+
+## Usage
+
+```py
+from dvclive import Live
+
+live = Live()
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_metric.md:12]: Not blocking so we can get this merged, but let's follow up on all these method snippets and use the context manager everywhere (and in this case, add `metrics.json` to the output).

File: content/docs/dvclive/api-reference/live/log_sklearn_plot.md [hunk 20]
```diff
@@ -63,19 +49,19 @@ plot.
...
 
 Example usage with `dvc plots`:
 
-```dvc
-$ dvc plots show 'dvclive/plots/calibration.json' \
+```cli
+$ dvc plots show 'dvclive/plots/sklearn/calibration.json' \
 -x prob_pred -y prob_true \
 --x-label 'Mean Predicted Probability' \
 --y-label 'Fraction of Positives' \
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_sklearn_plot.md:67]: ⚠️ **This is a blocker**: All the plots are missing in heroku except for precision recall

File: content/docs/dvclive/api-reference/live/make_summary.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
+# Live.make_summary()
+
+Serializes a summary of the logged metrics (`Live.summary`) to
+`Live.metrics_file`.
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/make_summary.md:4]: Not blocking:

```suggestion
Saves a summary of the latest metrics values (`Live.summary`) to
`dvclive/metrics.json`.
```

I think it's fine to either use `dvclive/metrics.json` or keep `Live.metrics_file` and link back to where it's explained.

File: content/docs/dvclive/api-reference/live/next_step.md [hunk 22]
```diff
@@ -14,33 +15,36 @@ from dvclive import Live
...
+and `Live.make_summary()`.
 
-<admon type="info">
+### Manual step updates
 
-Each `Live.next_step()` will call `Live.make_report()` internally by default
-(unless `report=None` is passed to `Live()`).
+If you want to decouple the `step` update from the rest of the calls, you can
+manually modify the `Live.step` property. This approach can be also used with
+custom `step` intervals:
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/next_step.md:36]:
```suggestion
If you want custom `step` intervals or don't want to call 
`Live.make_summary()` / `Live.make_report()`, you can
manually modify the `Live.step` property:
```

Non-blocking

File: content/docs/user-guide/experiment-management/checkpoints.md [hunk 23]
```diff
@@ -197,15 +197,15 @@ for i in range(1, EPOCHS+1):
     metrics = evaluate(model, x_test, y_test)
     for k, v in metrics.items():
         print('Epoch %s: %s=%s'%(i, k, v))
-+       dvclive.log(k, v)
++       dvclive.log_metric(k, v)
```
Comment (dberenbaum) [inline content/docs/user-guide/experiment-management/checkpoints.md:200]: Not blocking: should we go ahead and update `dvclive` to `live` here since that's the naming used everywhere else?

**Review by dberenbaum — CHANGES_REQUESTED** [review]
Amazing work to get this all done @daavoo!

I see two blockers:
1. Some of the sklearn plots images aren't showing up.
2. I think we are missing updates to `start/experiment-management/visualization.md`.

Hopefully I can take this over and fix those (assuming I have the plots images in the repo), and we can follow up on the rest later.

**Comment by dberenbaum** [comment]
Other issues to address:

1. https://github.com/iterative/dvc/issues/8531 (more of a core issue than docs)
2. https://github.com/iterative/example-get-started and https://github.com/iterative/example-dvc-experiments need to be updated

File: content/docs/start/experiment-management/visualization.md [hunk 24]
```diff
@@ -91,12 +91,12 @@ call.
...
 model.fit(
  ...
- callbacks=[DvcLiveCallback()],
+ callbacks=[DVCLiveCallback()],
  ...)
 ```
 
 With these two changes, the model metrics are automatically logged to
-`dvclive.json` and plotted in `training_metrics/index.html`:
+`training_metrics/metrics.json` and plotted in `training_metrics/report.html`:
```
Comment (daavoo) [inline content/docs/start/experiment-management/visualization.md:99]:
```suggestion
`dvclive/metrics.json` and plotted in `dvclive/report.html`:
```

**Comment by dberenbaum** [comment]
>     1. Some of the sklearn plots images aren't showing up.

@iterative/websites Can you take a look at the comment here?

https://github.com/iterative/dvc.org/pull/4085/files#r1014213049

It's showing for me locally but not in any of the heroku deployments. I tried clearing my browser data and using a private window, but no luck.

>     2. I think we are missing updates to `start/experiment-management/visualization.md`.

Fixed.

File: content/docs/dvclive/get-started.md (hunk 10 above)
Comment (daavoo) [reply content/docs/dvclive/get-started.md:90]: The thing for this report is that it was moved up to showcase value before entering in DVC integration, showing the other reports before the DVC integration could be misleading (for now)?

File: content/docs/dvclive/api-reference/index.md (hunk 13 above)
Comment (daavoo) [reply content/docs/dvclive/api-reference/index.md:33]: Makes sense.

**Comment by rogermparent** [comment]
I just pulled `main` into the branch, hopefully that helps things out. The dvc-based images can freak out without it.

**Comment by dberenbaum** [comment]
> I just pulled `main` into the branch, hopefully that helps things out. I just pushed it in here because we can't confirm without a deploy and we'd need to merge main at some point anyway. The dvc-based images can freak out if they're not updated.

Thanks @rogermparent! Still looks broken to me ☹️ . Any idea on whether we should expect it to work in prod?

**Comment by rogermparent** [comment]
> Any idea on whether we should expect it to work in prod?

I have no idea, maybe so if it's a simple cache issue. I'll spin up another fresh branch and see.

**Comment by rogermparent** [comment]
image
The copy worked!
My working theory is that either this PR's copy of the cache or this PR's S3 got desynchronized from main's, and it was expecting images to exist where they actually didn't.

As for if this will work on main, I'm pretty sure it _should_. New branches grab cache from `main`, so the build on `main` should be fine just like the new branch. Sorry for the inconvenience!

**Comment by dberenbaum** [comment]
Thanks @rogermparent! I guess we can keep an eye on it and revert if needed.

**Review by dberenbaum — APPROVED** [review]

File: content/docs/dvclive/api-reference/index.md (hunk 14 above)
Comment (jorgeorpinel) [reply content/docs/dvclive/api-reference/index.md:94]: > Why no mention of make_checkpoint()?

I'm also curious as to why this is not considered part of the https://dvc.org/doc/dvclive/api-reference#basic-workflow .

**Comment by jorgeorpinel** [comment]
> https://github.com/iterative/example-get-started and https://github.com/iterative/example-dvc-experiments need to be updated.

What's the plan on this? Esp. with the work on a new example repo

Thanks

**Comment by daavoo** [comment]
> What's the plan on this? Esp. with the work on a new example repo

Both are already updated and the new example repo is also using the latest version of DVCLive
